fem/restriction.hpp

```diff
@@ -216,7 +216,7 @@ class FaceRestriction : public Operator
 /// Operator that extracts Face degrees of freedom for H1 FiniteElementSpaces.
 /** Objects of this type are typically created and owned by FiniteElementSpace
     objects, see FiniteElementSpace::GetFaceRestriction(). */
-class H1FaceRestriction : public FaceRestriction
+class H1_ND_RT_FaceRestriction : public FaceRestriction
```

Comment:
I don't think that this monolithic class is a good design, the standard object oriented approach would be to factor common code in `FaceRestriction` and have `virtual` methods for what differs. The main reason for this is that monolithic classes tend to become unmaintainable and have poor extensibility. If performance is a concern then a classic base/derived template design pattern should solve this problem:
```c++
template < typename Derived >
class FaceRestrictionBase
{
   /// Common methods
};

class H1FaceRestriction: public FaceRestrictionBase< H1FaceRestriction >
{
   /// H1 specific code + methods used in FaceRestrictionBase
};
```

I know that `ElementRestriction` also shares this design, but I actually think that `ElementRestriction` should be split in H1, ND, RT, instead of mixing everything together in a monolithic class. I would also  remove `MultUnsigned`, `MultTransposeUnsigned` and put them in an `UnsignedRestriction` class ( containing `Mult` and `MultTranspose`, and use an `ElementRestriction` to build an `UnsignedRestriction` at no cost ).

Reply:
Hmm, I think there is a trade-off between having a more monolithic design (which may have maintainability drawbacks), and having a deeper, more complicated inheritance hierarchy where functionality is spread out across multiple layers of classes.

In particular, if you look at the diffs to `H1FaceRestriction`/`H1_ND_RT_FaceRestriction`, there is really only one change, which is incorporating the signs/orientations of the DOFs. The ND and RT code is a superset of the H1 code (the H1 implementation just never hits the "negative DOF" case). I pretty strong suspect specializing for each FE type will be less clear/result in more duplication. Also, the logic in `FiniteElementSpace::GetFaceRestriction` would have to be more complicated.

Reply:
Since @pazner is convinced that everything should behave the same between H1, ND, and RT, I guess this is fine.